src/program/latex/diagram/mod.rs

```diff
@@ -0,0 +1,151 @@
+    /// customizes how the diagram renders the circuit
+    pub(crate) settings: RenderSettings,
+    /// a BTreeMap of wires with the name of the wire as the key
+    pub(crate) circuit: BTreeMap<u64, Box<Wire>>,
+}
+
+impl Diagram {
+    /// Compares qubits from a single instruction associated with a column on
+    /// the circuit to all of the qubits used in the quil program. If a qubit
+    /// from the quil program is not found in the qubits in the single
+    /// instruction line, then an empty slot is added to that column on the
+    /// qubit wire of the circuit indicating a "do nothing" at that column.
+    ///
+    /// # Arguments
+    /// `qubits` - exposes the qubits used in the Program
+    /// `instruction` - exposes the qubits in a single Instruction
+    pub(crate) fn apply_empty(&mut self, qubits: &HashSet<Qubit>, gate: &Gate) {
+        qubits
+            .difference(&gate.qubits.iter().cloned().collect())
+            .filter_map(|q| match q {
+                Qubit::Fixed(index) => Some(index),
+                _ => None,
+            })
+            .for_each(|index| {
+                if let Some(wire) = self.circuit.get_mut(index) {
+                    wire.set_empty()
+                }
+            });
+    }
+
+    /// Applies a gate from an instruction to the associated wires in the
+    /// circuit of this diagram.
+    ///
+    /// # Arguments
+    /// `gate` - the Gate of the Instruction from `to_latex`.
+    pub(crate) fn apply_gate(&mut self, gate: &Gate) -> Result<(), LatexGenError> {
+        // set the parameters for each qubit in the instruction
+        for qubit in &gate.qubits {
+            if let Qubit::Fixed(qubit) = qubit {
+                if let Some(wire) = self.circuit.get_mut(qubit) {
+                    for expression in &gate.parameters {
+                        wire.set_param(expression, self.settings.texify_numerical_constants);
+                    }
+                }
+            }
+        }
+
+        // get the names of the qubits in the circuit before circuit is borrowed as mutable
+        let circuit_qubits: Vec<u64> = self.circuit.keys().cloned().collect();
+
+        // set gate for each qubit in the instruction
+        let last_qubit = gate.qubits.last().unwrap();
+        for (i, qubit) in gate.qubits.iter().enumerate() {
+            if let Qubit::Fixed(instruction_qubit) = qubit {
+                if let Some(wire) = self.circuit.get_mut(instruction_qubit) {
+                    // push the gate to the wire
+                    wire.gates.push(wire::T::try_from(gate.clone())?);
+
+                    // update the last gate to a control gate if the instruction has more than one qubit
+                    if gate.qubits.len() > 1 {
+                        if let wire::T::StdGate {
+                            name,
+                            dagger_count: _,
+                            ctrl_count,
+                        } = wire.gates.last().unwrap()
```

Comment:
Here, you push a gate onto the `Vec` and then pull it off and `unwrap()`. Why not just create it and use it directly?

Also, you discard information when you `T::try_from(gate.clone())`. You have a `gate`, which you know is a Gate - then you have a `T`, which might be a gate or something else. The Gate should be converted into something that is _more specific_ than a gate, not less. As from our last 1x1,  there should be a `struct` with the five fields derived from that gate. Then you can operate on that newer, more specific object directly.